### PR TITLE
CDD-2517: optional headers for CSPs

### PR DESCRIPTION
**Pull request by paulblakehmrc**

### CODE REVIEW

File: app/uk/gov/hmrc/customs/inventorylinking/export/controllers/HeaderValidator.scala [hunk 1]
```diff
@@ -80,5 +87,50 @@ class HeaderValidator @Inject()(logger: ExportsLogger) {
...
 
+    if (allowNone && maybeBadgeId.isEmpty) {
+      logger.info(s"$XBadgeIdentifierHeaderName header empty and allowed")
+      Right(None)
+    } else {
+      maybeBadgeId.filter(xBadgeIdentifierRegex.findFirstIn(_).nonEmpty).map(b => {
+        logger.info(s"$XBadgeIdentifierHeaderName header passed validation: $b")
+        Some(BadgeIdentifier(b))
+      }
+      ).toRight[ErrorResponse] {
```
Comment (glidester) [inline app/uk/gov/hmrc/customs/inventorylinking/export/controllers/HeaderValidator.scala]: I think 'map{ b => ...}' is clearer than 'map(b => {...})' save a set of brackets too.

File: app/uk/gov/hmrc/customs/inventorylinking/export/controllers/HeaderValidator.scala [hunk 2]
```diff
@@ -80,5 +87,50 @@ class HeaderValidator @Inject()(logger: ExportsLogger) {
...
+        errorResponseBadgeIdentifierHeaderMissing
+      }
+    }
+  }
+
+  private def validEori(eori: String) = InvalidEoriHeaderRegex.findFirstIn(eori).isEmpty
+
+  private def convertEmptyHeaderToNone(eori: Option[String]) = {
+    if (eori.isDefined && eori.get.trim.isEmpty) {
+      eori map (_.trim) filterNot (_.isEmpty)
```
Comment (glidester) [inline app/uk/gov/hmrc/customs/inventorylinking/export/controllers/HeaderValidator.scala]: Could this line not simply be replaced with 'None'?

File: public/api/conf/1.0/application.raml [hunk 3]
```diff
@@ -37,12 +37,12 @@ uses:
...
           example: application/xml; charset=UTF-8
         X-Badge-Identifier:
-            description: Required only for submissions via CSPs. Represents the end-user who is interacting with the subscribed application. Must be numeric or upper case letters with a minimum length of 6 and a maximum length of 12.
+            description: Represents the end-user who is interacting with the subscribed application. Must be numeric or upper case letters with a minimum length of 6 and a maximum length of 12. CSPs must have an X-Badge-Identifier or X-Submitter-Identifier.
             type: string
             required: false
             example: ABC123
         X-Submitter-Identifier:
-            description: A CSP user will populate this when submitting a transaction that has not originated with the CSP. The value must be an alphanumeric string with a maximum length of 17 char.
+            description: The EORI associated with the originating owner of the declaration. Must be numeric and upper case characters with a maximum length of 17 characters. CSPs must have an X-Badge-Identifier or X-Submitter-Identifier.
```
Comment (glidester) [inline public/api/conf/1.0/application.raml]: Is this still true? Thought this was mentioned at standup today as being changed.

File: public/api/conf/2.0/application.raml [hunk 4]
```diff
@@ -37,12 +37,12 @@ uses:
...
           example: application/xml; charset=UTF-8
         X-Badge-Identifier:
-            description: Required only for submissions via CSPs. Represents the end-user who is interacting with the subscribed application. Must be numeric or upper case letters with a minimum length of 6 and a maximum length of 12.
+            description: Represents the end-user who is interacting with the subscribed application. Must be numeric or upper case letters with a minimum length of 6 and a maximum length of 12. CSPs must have an X-Badge-Identifier or X-Submitter-Identifier.
             type: string
             required: false
             example: ABC123
         X-Submitter-Identifier:
-            description: A CSP user will populate this when submitting a transaction that has not originated with the CSP. The value must be an alphanumeric string with a maximum length of 17 char.
+            description: The EORI associated with the originating owner of the declaration. Must be numeric and upper case characters with a maximum length of 17 characters. CSPs must have an X-Badge-Identifier or X-Submitter-Identifier.
```
Comment (glidester) [inline public/api/conf/2.0/application.raml]: Is this still true? Thought this was mentioned at standup today as being changed.

File: app/uk/gov/hmrc/customs/inventorylinking/export/controllers/HeaderValidator.scala (hunk 1 above)
Comment (paulblakehmrc) [reply app/uk/gov/hmrc/customs/inventorylinking/export/controllers/HeaderValidator.scala]: Done.

File: public/api/conf/1.0/application.raml (hunk 3 above)
Comment (paulblakehmrc) [reply public/api/conf/1.0/application.raml]: I think for consistency across the API documentation we'll change it in all of them in one go or none of them.

File: public/api/conf/2.0/application.raml (hunk 4 above)
Comment (paulblakehmrc) [reply public/api/conf/2.0/application.raml]: I think for consistency across the API documentation we'll change it in all of them in one go or none of them.

File: app/uk/gov/hmrc/customs/inventorylinking/export/controllers/HeaderValidator.scala (hunk 2 above)
Comment (paulblakehmrc) [reply app/uk/gov/hmrc/customs/inventorylinking/export/controllers/HeaderValidator.scala]: I think it can. Done.

**Pull request merged**